rubicon/java/android_events.py

```diff
@@ -168,45 +193,125 @@ def _run_scheduled_callbacks_then_restart_loop(self):
                 handle._run()
         handle = None  # Needed to break cycles when an exception occurs.
 
-        self._android_call_later(
-            self._runnables["_run_event_loop_once_cooperatively"], 0
-        )
+        # End code borrowed from CPython, within this method.
+        self.enqueue_android_wakeup_for_delayed_tasks()
 
-    def run_forever_cooperatively(self):
-        # Based heavily on `BaseEventLoop.run_forever()` in CPython.
-        if self.is_running():
-            raise RuntimeError("Refusing to start since loop is already running.")
-        if self._closed:
-            raise RuntimeError("Event loop is closed. Create a new object.")
-        self._set_coroutine_origin_tracking(self._debug)
-        self._thread_id = threading.get_ident()
 
-        self._old_agen_hooks = sys.get_asyncgen_hooks()
-        sys.set_asyncgen_hooks(
-            firstiter=self._asyncgen_firstiter_hook,
-            finalizer=self._asyncgen_finalizer_hook,
+class _SelectorMinusSelect(selectors.PollSelector):
+    # This class removes the `select()` method from PollSelector, purely as
+    # a safety mechanism. On Android, this would be an error -- it would result
+    # in the app freezing, triggering an App Not Responding pop-up from the
+    # platform, and the user killing the app.
+    #
+    # Instead, the AndroidEventLoop cooperates with the native Android event
+    # loop to be woken up to get work done as needed.
+    def select(self, *args, **kwargs):
+        raise NotImplementedError(
+            "_SelectorMinusSelect refuses to select(); see comments."
```

Comment:
Would it would be woefully naive to think that making select() a no-op might simplify other pieces of the implementation? It wouldn't surprise me if the answer is no - but just in case...

Reply:
If we do that, then we might end up looping on it forever :)

The way the real event loop works is:

```
while True:
    file_descriptors_ready = select(all_file_descriptors_we_care_about, timeout)
    process_file_descriptors(file_descriptors_ready)
```

I'm exaggerating on the variable names and function names, but if we make `select` a no-op, it becomes a forever `while True` loop.

Reply:
Right - that makes sense.